# Proyecto integrador para el Módulo de analista de Datos - ISPC

#### Módulo: Analista de datos (Estádistica y Exploración de datos, Ciencia de Datos y Procesamiento de Datos)
#### Alumno: PALOMEQUE, Jonathan Manuel
#### Carrera: Tecnicatura Superior en Ciencia de datos e IA
#### Institución: Instituto Superior Politécnico Córdoba (ISPC)
#### Año: 2024

## Desarrollo del proyecto: 

### I) Importar Pandas, Numpy y Matplotlib:

In [215]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

### II) Cargar DataSet y crear el Dataframe:

In [216]:
dataSet = pd.read_csv('Ventas-Original.csv')

dataframe = pd.DataFrame(dataSet)

### III) Descripcion del DataSet:

**1- Informacion General sobre el dataframe:**

In [217]:
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12014 entries, 0 to 12013
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   pais              12004 non-null  object 
 1   metodo_orden      12008 non-null  object 
 2   tipo_comercio     12005 non-null  object 
 3   linea             12004 non-null  object 
 4   tipo_producto     12008 non-null  object 
 5   producto          12005 non-null  object 
 6   año               12009 non-null  float64
 7   cuatrimestre      12014 non-null  object 
 8   ventas            12009 non-null  object 
 9   cantidad          12009 non-null  float64
 10  utilidad_bruta_p  12010 non-null  object 
 11  utilidad_bruta    12007 non-null  object 
 12  costos            12009 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.2+ MB
None


*El resultado brinda informacion de la cantidad de columnas, sus nombres, tipoa de datos y que hay 12.014 filas y 13 columnas. No hay valores Nulos en las columnas*

**2- Imprimir los primeros  10 registros del dataframe:**

In [218]:
print(dataframe.head(10))

        pais   metodo_orden      tipo_comercio     linea tipo_producto  \
0  Argentina  Representante          Mayorista  Mascotas      Juguetes   
1  Argentina           Mail      De proximidad  Mascotas      Juguetes   
2  Argentina       Teléfono  Grandes Almacenes  Infantil       Calzado   
3  Argentina        Website      Especializado    Hombre       Calzado   
4  Argentina        Website      Especializado   Adultos    Vestimenta   
5  Argentina        Website          Mayorista     Mujer        Lentes   
6  Argentina       Teléfono      Especializado     Mujer    Vestimenta   
7  Argentina  Representante       Hipermercado    Hombre        Lentes   
8  Argentina  Representante  Grandes Almacenes  Mascotas      Juguetes   
9  Argentina  Representante          Mayorista    Hombre       Calzado   

                 producto     año cuatrimestre       ventas  cantidad  \
0        Anitibyx armatus  2019.0      C1 2019  $155.040,02   15562.0   
1      Sarcorhamphus papa  2019.0      

**3- Cantidad de registros y columnas:**

In [219]:
cant_filas, cant_columnas = dataframe.shape

print("""
La Cantidad de registros son: {}
La Cantidad de columnas son: {}""".format(cant_filas, cant_columnas))


La Cantidad de registros son: 12014
La Cantidad de columnas son: 13


**4- Cantidad de elementos que NO son nulos por columnas:**

In [220]:
dataframe.count()

pais                12004
metodo_orden        12008
tipo_comercio       12005
linea               12004
tipo_producto       12008
producto            12005
año                 12009
cuatrimestre        12014
ventas              12009
cantidad            12009
utilidad_bruta_p    12010
utilidad_bruta      12007
costos              12009
dtype: int64

**5-Tipos de Datos:**

In [221]:
tipos_datos = dataframe.dtypes
print(tipos_datos)

pais                 object
metodo_orden         object
tipo_comercio        object
linea                object
tipo_producto        object
producto             object
año                 float64
cuatrimestre         object
ventas               object
cantidad            float64
utilidad_bruta_p     object
utilidad_bruta       object
costos               object
dtype: object


*Verifico que hay columnas cuyos datos deberian ser del tipo entero o flotante, los cuales actualmente son del tipo objeto, por lo que deben ser modificados para poder realizar el analisis de los mismos*

### IV) Limpieza y Procesamiento de datos:

**1- Limpieza de celdas vacias:**

*Utilizo el metodo  dropna() para eliminar las filas que contengan celdas vacias, como deseo conservar el DataFrame original, utilizo el argumento inplace=True*

In [222]:
dataframe.dropna(inplace = True)

*Verifico la cantidad de registros  resultantes luego de la limpieza:*

In [223]:
ncant_filas, ncant_columnas = dataframe.shape
print(f"Cantidad de registros: {ncant_filas}. Cantidad de columnas: {ncant_columnas}.")

Cantidad de registros: 11952. Cantidad de columnas: 13.


*De un total original de 12014 filas, quedaron 11952, lo que nos indica que habia 62 filas con celdas vacias*

**2- Datos en Formato incorrecto:**

Es necesario cambiar el tipo de dato de las columnas: *ventas, utilidad_bruta,  costos y utilidad_bruta_p*, del tipo objeto al tipo flotante, ya que verifico que los registros de las columnas ventas, utilidad_bruta y costos, originalmente tiene la siguiente sintaxis: "$155.040,02" y necesito que queden de la siguiente forma: "155040.02" , para poder obtener las medidas de tendencias centrales (media, mediana y desviacion estandar) asi como para realizar acciones de conteo. 

Para hacerlo seguiré los siguientes pasos:
- eliminar los signos "$"
- eliminar as ","
- reemplazar los "." por "," (Ya que sino, no puedo cambiar el tipo de dato)

**A- Reemplazar los signos "$" de las columnas: ventas, 'utilidad_bruta' y 'costos'**

In [224]:
dataframe['ventas'] = dataframe['ventas'].str.replace('$', '')
dataframe['utilidad_bruta'] = dataframe['utilidad_bruta'].str.replace('$', '')
dataframe['costos'] = dataframe['costos'].str.replace('$', '')

**B- Reemplazar los signos "." de las columnas: ventas, 'utilidad_bruta', 'costos' y 'cantidad'**

In [225]:
dataframe['ventas'] = dataframe['ventas'].str.replace('.', '')
dataframe['utilidad_bruta'] = dataframe['utilidad_bruta'].str.replace('.', '')
dataframe['costos'] = dataframe['costos'].str.replace('.', '')

**C - Reemplazar los signos "," por "." de las columnas:ventas, , 'utilidad_bruta', 'costos' y 'cantidad'**

In [226]:
dataframe['ventas'] = dataframe['ventas'].str.replace(',', '.')
dataframe['utilidad_bruta'] = dataframe['utilidad_bruta'].str.replace(',', '.')
dataframe['costos'] = dataframe['costos'].str.replace(',', '.')

**D- Reemplazar los signos "%" por "" y "." por "," de la columna "utilidad_bruta_p"**

In [227]:
dataframe['utilidad_bruta_p'] = dataframe['utilidad_bruta_p'].str.replace('%','')
dataframe['utilidad_bruta_p'] = dataframe['utilidad_bruta_p'].str.replace(',','.')

**E- Ahora si, puedo convertir cambiar el tipo de dato a float:**

In [228]:
dataframe[['ventas', 'utilidad_bruta', 'costos', 'utilidad_bruta_p']] = dataframe[['ventas', 'utilidad_bruta', 'costos', 'utilidad_bruta_p']].astype(float)

**F- Cambiar el tipo de dato de la columna año al tipo entero:**

In [229]:
dataframe['año'] = dataframe['año'].astype(int)

**G-Tipos de datos finales:**

In [230]:
tipos_datos = dataframe.dtypes
print(tipos_datos)

pais                 object
metodo_orden         object
tipo_comercio        object
linea                object
tipo_producto        object
producto             object
año                   int32
cuatrimestre         object
ventas              float64
cantidad            float64
utilidad_bruta_p    float64
utilidad_bruta      float64
costos              float64
dtype: object


**3- Limpieza de filas duplicadas:**

**A- Identificar las filas duplicadas:**

*Uso el metodo duplicated() para identificar duplicados, como este metodo nos devuelve True si las filas estan duplicadas, no es practico de visualizar cuando son muchos registros, por lo que primero guardo el resultado en un nuevo dataset (filasRegistro), para luego recorrer cada celda  y guardar en un array el indice de aquellas que son  True (duplicados), asi poder llevar un control de cuales son las duplicadas.*

In [231]:
filasRegistro = dataframe.duplicated()
filasDuplicadas = []

for x in filasRegistro.index:
  if filasRegistro.loc[x] == True:
    filasDuplicadas.append(x)

if len(filasDuplicadas) == 0:
    print('No hay filas duplicadas')
else:
    print('Hay un total de {} filas duplicadas, las cuales son: {}'.format(len(filasDuplicadas), filasDuplicadas))

Hay un total de 14 filas duplicadas, las cuales son: [11, 1973, 2001, 2009, 3499, 7091, 7871, 7908, 9269, 9301, 9677, 9679, 11750, 11778]


**B - Eliminar filas duplicadas:**

In [232]:
dataframe.drop_duplicates(inplace = True)

*Verifico la cantidad de registros resultantes luego de la limpieza:*

In [233]:
cant_filas, cant_columnas = dataframe.shape
print(f"Cantidad de registros resultantes: {cant_filas}. Cantidad de columnas: {cant_columnas}")

Cantidad de registros resultantes: 11938. Cantidad de columnas: 13


*De un total original de 12.014 filas, pasamos a 11.952 y ahora nos quedan 11.938 lo que nos indica que habia 14 filas con celdas duplicadas*

**4- Limpieza de información incorrecta:**

**A- Verificar registros que sean igual o menor a 0 en las columnas: año, ventas, cantidad, utilidad_bruta, utilidad_bruta_p y costos**

*Utilizare filtros, para detectas estas filas*

In [234]:
menorIgualaCeroAño = dataframe[dataframe['año'] <= 0]
menorIgualaCeroVentas = dataframe[dataframe['ventas'] <= 0]
menorIgualaCeroCantidad = dataframe[dataframe['cantidad'] <= 0]
menorIgualaCeroUBP =dataframe[dataframe['utilidad_bruta_p'] <= 0]
menorIgualaCeroUB =dataframe[dataframe['utilidad_bruta'] <= 0]
menorIgualaCeroCostos =dataframe[dataframe['costos'] <= 0]

print('''Cantidad de elementos menores o iguales a cero, por columna:
- Año: {}
- Ventas: {}
- Cantidad: {}
- Utilidad Bruta P: {}
- Utilidad Bruta: {}
- Costos:{}
'''.format(menorIgualaCeroAño.size, menorIgualaCeroVentas.size, menorIgualaCeroCantidad.size, 
           menorIgualaCeroUBP.size, menorIgualaCeroUB.size, menorIgualaCeroCostos.size))

Cantidad de elementos menores o iguales a cero, por columna:
- Año: 0
- Ventas: 78
- Cantidad: 65
- Utilidad Bruta P: 0
- Utilidad Bruta: 0
- Costos:0



*Del analisis verifico que en las columnas  cantidad y ventas, hay valores que son iguales o menores a Cero, lo cual es un error*

**B- Borrar los registros  de ventas que sean igual a ceros**

*En esta columna decido borrar los registros, y conservar aquellos donde el valor de ventas sea mayor a cero, ya que no tengo un valor de referencia*

In [235]:
dataframe.drop(dataframe[dataframe['ventas']<=0].index, inplace=True)

*Se verifica si hay registros menores o igual a cero en la columna ventas, paraa verificar que los cambios funcionaron correctamente*

In [236]:
dataframe[dataframe['ventas'] <= 0]

Empty DataFrame
Columns: [pais, metodo_orden, tipo_comercio, linea, tipo_producto, producto, año, cuatrimestre, ventas, cantidad, utilidad_bruta_p, utilidad_bruta, costos]
Index: []

Se verifica que ya no hay registros iguales o menores a cero para la columna 

**C-Reemplazar los valores en negativo por su equivalente, para la columna cantidad**

*En este caso si hay un valor de referencia, lo que esta mal es que estan en negativo*

In [237]:
dataframe.loc[dataframe["cantidad"] <= 0, "cantidad"] *= -1

*Verifico si quedaron valores menores o iguales a cero para la columna cantidad:*

In [238]:
dataframe[dataframe['cantidad']<=0]

Empty DataFrame
Columns: [pais, metodo_orden, tipo_comercio, linea, tipo_producto, producto, año, cuatrimestre, ventas, cantidad, utilidad_bruta_p, utilidad_bruta, costos]
Index: []

*ya no hay valores menores o iguales a cero para la columna cantidad*

**D- Verifico como el estado del dataFrame:**

*Se verifica que quedaron un total de 11.932 registros*

In [241]:
print(dataframe.info())

<class 'pandas.core.frame.DataFrame'>
Index: 11932 entries, 0 to 12013
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   pais              11932 non-null  object 
 1   metodo_orden      11932 non-null  object 
 2   tipo_comercio     11932 non-null  object 
 3   linea             11932 non-null  object 
 4   tipo_producto     11932 non-null  object 
 5   producto          11932 non-null  object 
 6   año               11932 non-null  int32  
 7   cuatrimestre      11932 non-null  object 
 8   ventas            11932 non-null  float64
 9   cantidad          11932 non-null  float64
 10  utilidad_bruta_p  11932 non-null  float64
 11  utilidad_bruta    11932 non-null  float64
 12  costos            11932 non-null  float64
dtypes: float64(5), int32(1), object(7)
memory usage: 1.2+ MB
None


### V) Estadísticas:

**1- Informacion estadistica del dataset:**

In [242]:
infoEstadistica = dataframe.describe()
infoEstadistica.round(2)

año     ventas  cantidad  utilidad_bruta_p  utilidad_bruta  \
count   11932.00   11932.00  11932.00          11932.00        11932.00   
mean     2184.41  138031.58   9008.08             49.90        68883.99   
std      5229.12   27287.11   4901.00             14.78        24824.31   
min      2019.00   90530.25    526.00             24.58        22407.08   
25%      2019.00  114488.61   4753.75             37.16        49131.29   
50%      2020.00  137799.99   8999.50             49.72        66375.19   
75%      2021.00  161723.40  13244.50             62.90        85775.83   
max    201900.00  185644.24  17521.00             75.40       138447.51   

          costos  
count   11932.00  
mean    69147.60  
std     24831.64  
min     22339.56  
25%     49478.43  
50%     66511.94  
75%     86448.51  
max    139072.81

**2- Media:**

In [243]:
# media de Ventas:
media_ventas = dataframe['ventas'].mean()

# media de Cantidad de productos:
media_cant_productos = dataframe['cantidad'].mean()

# media de Utilidad Bruta:
media_utilidad_bruta = dataframe['utilidad_bruta'].mean()

# media de costos:
media_costos =  dataframe['costos'].mean()


print("""
La media de ventas es: ${} 
La media de la cantidad de productos es: {} 
La media de utilidades brutas es: ${} 
La media de costos es: ${}
""".format(media_ventas.round(2), media_cant_productos.round(2), media_utilidad_bruta.round(2), media_costos.round(2)))


La media de ventas es: $138031.58 
La media de la cantidad de productos es: 9008.08 
La media de utilidades brutas es: $68883.99 
La media de costos es: $69147.6



**3- Mediana:**

In [244]:
# mediana de Ventas:
mediana_ventas = dataframe['ventas'].median()

# mediana de Cantidad de productos:
mediana_cant_productos = dataframe['cantidad'].median()

# mediana de Utilidad Bruta:
mediana_utilidad_bruta = dataframe['utilidad_bruta'].median()

# mediana de costos:
mediana_costos =  dataframe['costos'].median()


print("""
La mediana de ventas es: ${} 
La mediana de la cantidad de productos es: {} 
La mediana de utilidades brutas es: ${} 
La mediana de costos es: ${}
""".format(mediana_ventas, mediana_cant_productos, mediana_utilidad_bruta, mediana_costos))


La mediana de ventas es: $137799.99 
La mediana de la cantidad de productos es: 8999.5 
La mediana de utilidades brutas es: $66375.19 
La mediana de costos es: $66511.94



**4- Desviacion Estandar:**

In [245]:
# Desviacion estándar en ventas:
desviacion_ventas = dataframe['ventas'].std()

#Desviación estándar de Cantidad de Productos:
desviacion_cant_productos = dataframe['cantidad'].std()

#Desviación estándar de utilidad bruta:
desviacion_utilidad_bruta = dataframe['utilidad_bruta'].std()

#Desviacion estándar de costos:
desviacion_costos = dataframe['costos'].std()

print("""
La desviación estándar de ventas es: ${}
La Desviación estándar de cantidad de productos es: {}
La desviación estándar de utilidad bruta es: {}
La desviacion estandar de costos es: {}""".format(desviacion_ventas, desviacion_cant_productos, desviacion_utilidad_bruta, desviacion_costos))


La desviación estándar de ventas es: $27287.107331194882
La Desviación estándar de cantidad de productos es: 4900.998441872444
La desviación estándar de utilidad bruta es: 24824.309456247614
La desviacion estandar de costos es: 24831.638724164717
